### PR TITLE
Travis: Build tests on Ubuntu 18.04 with docker

### PR DESCRIPTION
**Pull request by ken2812221**
Compile and run tests on Ubuntu 18.04 docker.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
I can't see any downside, so Concept ACK.

**Comment by promag** [comment]
Can you elaborate the motivation behind this change?

File: .travis.yml [hunk 1]
```diff
@@ -19,58 +19,62 @@ env:
...
+    - HOST=x86_64-apple-darwin11 PACKAGES="cmake imagemagick libcap-dev librsvg2-bin libz-dev libbz2-dev libtiff-tools python-dev python3-setuptools-git" BITCOIN_CONFIG="--enable-gui --enable-reduce-exports --enable-werror" OSX_SDK=10.11 GOAL="deploy"
 
 before_install:
     - export PATH=$(echo $PATH | tr ':' "\n" | sed '/\/opt\/python/d' | tr "\n" ":" | sed "s|::|:|g")
 install:
-    - if [ -n "$DPKG_ADD_ARCH" ]; then sudo dpkg --add-architecture "$DPKG_ADD_ARCH" ; fi
-    - if [ -n "$PACKAGES" ]; then travis_retry sudo apt-get update; fi
-    - if [ -n "$PACKAGES" ]; then travis_retry sudo apt-get install --no-install-recommends --no-upgrade -qq $PACKAGES; fi
+    - env | grep -vE '^(LC_.*|LANG)=' > /tmp/env
+    - DOCKER_ID=$(docker run --privileged -idtv $TRAVIS_BUILD_DIR:$TRAVIS_BUILD_DIR -v $HOME/.ccache:$HOME/.ccache -w $TRAVIS_BUILD_DIR --env-file /tmp/env  ubuntu:18.04)
```
Comment (promag) [inline .travis.yml]: Remove double space.

**Review by promag — COMMENTED** [review]
Concept ACK.

Maybe it should also cache docker layers? Similar to http://circleci.com/docs/1.0/docker/#caching-docker-layers.

Edit: seems to take around 30 to 40 seconds in pulling images and installing extra packages.

**Comment by MarcoFalke** [comment]
@promag I presume the motivation is that gitian will (soon) use bionic, so having a similar configuration run on travis isn't that bad. Also, it untangles us a bit from the travis environments and changes, as we use the vanilla bionic docker. So reproducing issues could potentially be easier or harder because it uses docker ...

I have no strong opinion based on that, but since the new qt "requires" bionic for windows cross builds, I am slightly in favor of this change.

**Comment by ken2812221** [comment]
>Can you elaborate the motivation behind this change?

We will bump our gitian build system to Ubuntu 18.04 since 0.17, so I think we might need to have the same build environment as the gitian build. Also, once we've merged the back-compat code, we could run the executable both on 14.04 and 18.04 to test if the back-compat code would work.

**Comment by jamesob** [comment]
This will likely fix the build on https://github.com/bitcoin/bitcoin/pull/12873 by (inadvertently) using a more recent version of SSL that doesn't cause the deadlock warning system to flare up.

The job duration times don't seem adversely affected by this change.

Concept ACK

File: .travis.yml (hunk 1 above)
Comment (ken2812221) [reply .travis.yml]: Done.

**Comment by MarcoFalke** [comment]
Please remove the [WIP] from the commit title when this is ready for review.

File: .travis.yml [hunk 2]
```diff
@@ -19,58 +19,62 @@ env:
...
+    - HOST=x86_64-apple-darwin11 PACKAGES="cmake imagemagick libcap-dev librsvg2-bin libz-dev libbz2-dev libtiff-tools python-dev python3-setuptools-git" BITCOIN_CONFIG="--enable-gui --enable-reduce-exports --enable-werror" OSX_SDK=10.11 GOAL="deploy"
 
 before_install:
     - export PATH=$(echo $PATH | tr ':' "\n" | sed '/\/opt\/python/d' | tr "\n" ":" | sed "s|::|:|g")
 install:
-    - if [ -n "$DPKG_ADD_ARCH" ]; then sudo dpkg --add-architecture "$DPKG_ADD_ARCH" ; fi
-    - if [ -n "$PACKAGES" ]; then travis_retry sudo apt-get update; fi
-    - if [ -n "$PACKAGES" ]; then travis_retry sudo apt-get install --no-install-recommends --no-upgrade -qq $PACKAGES; fi
+    - env | grep -vE '^(LC_.*|LANG)=' > /tmp/env
+    - DOCKER_ID=$(docker run --privileged -idtv $TRAVIS_BUILD_DIR:$TRAVIS_BUILD_DIR -v $HOME/.ccache:$HOME/.ccache -w $TRAVIS_BUILD_DIR --env-file /tmp/env ubuntu:18.04)
```
Comment (jamesob) [inline .travis.yml]: Small thing, but could we add 
```yaml
- DOCKER_EXEC="docker exec ${DOCKER_ID}"
```
to cut down on some of the boilerplate below?

Comment (ken2812221) [reply .travis.yml]: Sure

**Comment by jamesob** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/13215/commits/9e7924c0cb0995f8df6cca5d04b576e5b56038a9

**Comment by ken2812221** [comment]
> Maybe it should also cache docker layers? Similar to http://circleci.com/docs/1.0/docker/#caching-docker-layers.

I've tested this, `docker save` took more time than `apt install`.

**Comment by promag** [comment]
@ken2812221 indeed in travis it's fast.

utACK 9e7924c.

File: .travis.yml [hunk 3]
```diff
@@ -19,58 +19,63 @@ env:
...
-    - HOST=x86_64-apple-darwin11 PACKAGES="cmake imagemagick libcap-dev librsvg2-bin libz-dev libbz2-dev libtiff-tools python-dev" BITCOIN_CONFIG="--enable-gui --enable-reduce-exports --enable-werror" OSX_SDK=10.11 GOAL="deploy"
+    - HOST=x86_64-apple-darwin11 PACKAGES="cmake imagemagick libcap-dev librsvg2-bin libz-dev libbz2-dev libtiff-tools python-dev python3-setuptools-git" BITCOIN_CONFIG="--enable-gui --enable-reduce-exports --enable-werror" OSX_SDK=10.11 GOAL="deploy"
 
 before_install:
     - export PATH=$(echo $PATH | tr ':' "\n" | sed '/\/opt\/python/d' | tr "\n" ":" | sed "s|::|:|g")
 install:
-    - if [ -n "$DPKG_ADD_ARCH" ]; then sudo dpkg --add-architecture "$DPKG_ADD_ARCH" ; fi
-    - if [ -n "$PACKAGES" ]; then travis_retry sudo apt-get update; fi
-    - if [ -n "$PACKAGES" ]; then travis_retry sudo apt-get install --no-install-recommends --no-upgrade -qq $PACKAGES; fi
+    - env | grep -vE '^(LC_.*|LANG)=' > /tmp/env
```
Comment (MarcoFalke) [inline .travis.yml]: nit: I'd prefer if we only write the environment variables that are specified in the yaml (global env and job env). Currently this also exports a lot of travis specific junk, which shouldn't be necessary. No strong opinion, though, as I am not aware how to achieve this.

File: .travis.yml [hunk 4]
```diff
@@ -19,58 +19,63 @@ env:
...
+    - DOCKER_ID=$(docker run --privileged -idtv $TRAVIS_BUILD_DIR:$TRAVIS_BUILD_DIR -v $HOME/.ccache:$HOME/.ccache -w $TRAVIS_BUILD_DIR --env-file /tmp/env ubuntu:18.04)
```
Comment (MarcoFalke) [inline .travis.yml]: nits:
* Please replace the `-v` with the more verbose `--mount type=bind,src=....,dst=....`, which is suggested in the docker docs
* You overwrite the `CCACHE_DIR` some lines above, but don't use that value here. Any reason for that?
* Please replace the privliledged flag with just `--cap-add SYS_ADMIN` and add a comment why this is required or make it specific to the windows build.

File: .travis.yml [hunk 5]
```diff
@@ -19,58 +19,63 @@ env:
...
+    - DOCKER_EXEC="docker exec $DOCKER_ID"
+    - if [ -n "$DPKG_ADD_ARCH" ]; then $DOCKER_EXEC dpkg --add-architecture "$DPKG_ADD_ARCH" ; fi
+    - travis_retry $DOCKER_EXEC apt-get update
+    - travis_retry $DOCKER_EXEC apt-get install --no-install-recommends --no-upgrade -qq $PACKAGES $DOCKER_PACKAGES
```
Comment (MarcoFalke) [inline .travis.yml]: nit: Can use `apt` instead of `apt-get`

File: .travis.yml [hunk 6]
```diff
@@ -19,58 +19,63 @@ env:
...
-    - make distdir VERSION=$HOST
+    - $DOCKER_EXEC bash -c "cd build && ../configure --cache-file=config.cache $BITCOIN_CONFIG_ALL $BITCOIN_CONFIG" || ( cat config.log && false)
+    - $DOCKER_EXEC bash -c "cd build && make distdir VERSION=$HOST"
     - cd bitcoin-$HOST
-    - ./configure --cache-file=../config.cache $BITCOIN_CONFIG_ALL $BITCOIN_CONFIG || ( cat config.log && false)
-    - make $MAKEJOBS $GOAL || ( echo "Build failure. Verbose build follows." && make $GOAL V=1 ; false )
-    - export LD_LIBRARY_PATH=$TRAVIS_BUILD_DIR/depends/$HOST/lib
-    - if [ "$RUN_TESTS" = "true" ]; then travis_wait 50 make $MAKEJOBS check VERBOSE=1; fi
+    - $DOCKER_EXEC bash -c "cd build/bitcoin-$HOST && ./configure --cache-file=../config.cache $BITCOIN_CONFIG_ALL $BITCOIN_CONFIG" || ( cat config.log && false)
+    - $DOCKER_EXEC bash -c "cd build/bitcoin-$HOST && make $MAKEJOBS $GOAL" || ( echo "Build failure. Verbose build follows." && make $GOAL V=1 ; false )
```
Comment (MarcoFalke) [inline .travis.yml]: This fails, since it is executed on the host, not via DOCKER_EXEC.

**Comment by MarcoFalke** [comment]
Tested ACK 9e7924c0cb0995f8df6cca5d04b576e5b56038a9 (after fixing the failure)

Also left a couple of nits in the process, you can safely ignore them.

File: .travis.yml [hunk 7]
```diff
@@ -19,58 +19,64 @@ env:
...
-    - HOST=x86_64-apple-darwin11 PACKAGES="cmake imagemagick libcap-dev librsvg2-bin libz-dev libbz2-dev libtiff-tools python-dev" BITCOIN_CONFIG="--enable-gui --enable-reduce-exports --enable-werror" OSX_SDK=10.11 GOAL="deploy"
+    - HOST=x86_64-apple-darwin11 PACKAGES="cmake imagemagick libcap-dev librsvg2-bin libz-dev libbz2-dev libtiff-tools python-dev python3-setuptools-git" BITCOIN_CONFIG="--enable-gui --enable-reduce-exports --enable-werror" OSX_SDK=10.11 GOAL="deploy"
 
 before_install:
     - export PATH=$(echo $PATH | tr ':' "\n" | sed '/\/opt\/python/d' | tr "\n" ":" | sed "s|::|:|g")
 install:
-    - if [ -n "$DPKG_ADD_ARCH" ]; then sudo dpkg --add-architecture "$DPKG_ADD_ARCH" ; fi
-    - if [ -n "$PACKAGES" ]; then travis_retry sudo apt-get update; fi
-    - if [ -n "$PACKAGES" ]; then travis_retry sudo apt-get install --no-install-recommends --no-upgrade -qq $PACKAGES; fi
+    - env | grep -E '^(CCACHE_|WINEDEBUG|DISPLAY)' > /tmp/env
```
Comment (MarcoFalke) [inline .travis.yml]: I assume `BOOST_TEST_RANDOM` and `DISPLAY` needs to be passed down as well?

**Comment by MarcoFalke** [comment]
re-ACK 9e7924c

**Comment by theuni** [comment]
Playing around with this now.

File: .travis.yml [hunk 8]
```diff
@@ -19,58 +19,63 @@ env:
...
+    - HOST=x86_64-apple-darwin11 PACKAGES="cmake imagemagick libcap-dev librsvg2-bin libz-dev libbz2-dev libtiff-tools python-dev python3-setuptools-git" BITCOIN_CONFIG="--enable-gui --enable-reduce-exports --enable-werror" OSX_SDK=10.11 GOAL="deploy"
 
 before_install:
     - export PATH=$(echo $PATH | tr ':' "\n" | sed '/\/opt\/python/d' | tr "\n" ":" | sed "s|::|:|g")
 install:
-    - if [ -n "$DPKG_ADD_ARCH" ]; then sudo dpkg --add-architecture "$DPKG_ADD_ARCH" ; fi
-    - if [ -n "$PACKAGES" ]; then travis_retry sudo apt-get update; fi
-    - if [ -n "$PACKAGES" ]; then travis_retry sudo apt-get install --no-install-recommends --no-upgrade -qq $PACKAGES; fi
+    - env | grep -E '^(CCACHE_|WINEDEBUG|DISPLAY|BOOST_TEST_RANDOM)' | tee /tmp/env
+    - if [[ $HOST = *-mingw32 ]]; then DOCKER_ADMIN="--cap-add SYS_ADMIN"; fi
```
Comment (theuni) [inline .travis.yml]: Why is this needed?

Comment (ken2812221) [reply .travis.yml]: In order to set wine-binfmt

File: .travis.yml [hunk 9]
```diff
@@ -19,58 +19,63 @@ env:
...
+    - DOCKER_ID=$(docker run $DOCKER_ADMIN -idt --mount type=bind,src=$TRAVIS_BUILD_DIR,dst=$TRAVIS_BUILD_DIR --mount type=bind,src=$CCACHE_DIR,dst=$CCACHE_DIR -w $TRAVIS_BUILD_DIR --env-file /tmp/env ubuntu:18.04)
+    - DOCKER_EXEC="docker exec $DOCKER_ID"
+    - if [ -n "$DPKG_ADD_ARCH" ]; then $DOCKER_EXEC dpkg --add-architecture "$DPKG_ADD_ARCH" ; fi
```
Comment (theuni) [inline .travis.yml]: Is this still needed? IIRC it was an artifact of needing a non-default wine.

Comment (ken2812221) [reply .travis.yml]: Yes, this is needed. `wine32` is only available on i386 platform. Here is the link: https://packages.ubuntu.com/bionic/wine32

File: .travis.yml [hunk 10]
```diff
@@ -19,58 +19,63 @@ env:
...
+    - $DOCKER_EXEC bash -c "cd build && ../configure --cache-file=config.cache $BITCOIN_CONFIG_ALL $BITCOIN_CONFIG" || ( cat config.log && false)
+    - $DOCKER_EXEC bash -c "cd build && make distdir VERSION=$HOST"
     - cd bitcoin-$HOST
-    - ./configure --cache-file=../config.cache $BITCOIN_CONFIG_ALL $BITCOIN_CONFIG || ( cat config.log && false)
-    - make $MAKEJOBS $GOAL || ( echo "Build failure. Verbose build follows." && make $GOAL V=1 ; false )
-    - export LD_LIBRARY_PATH=$TRAVIS_BUILD_DIR/depends/$HOST/lib
-    - if [ "$RUN_TESTS" = "true" ]; then travis_wait 50 make $MAKEJOBS check VERBOSE=1; fi
+    - $DOCKER_EXEC bash -c "cd build/bitcoin-$HOST && ./configure --cache-file=../config.cache $BITCOIN_CONFIG_ALL $BITCOIN_CONFIG" || ( cat config.log && false)
+    - $DOCKER_EXEC bash -c "cd build/bitcoin-$HOST && make $MAKEJOBS $GOAL" || ( echo "Build failure. Verbose build follows." && $DOCKER_EXEC bash -c "cd build/bitcoin-$HOST && make $GOAL V=1" ; false )
+    - if [ "$RUN_TESTS" = "true" ]; then $DOCKER_EXEC bash -c "cd build/bitcoin-$HOST && LD_LIBRARY_PATH=$TRAVIS_BUILD_DIR/depends/$HOST/lib make $MAKEJOBS check VERBOSE=1"; fi
```
Comment (theuni) [inline .travis.yml]: This dropped the "travis_wait 50", but I guess we're ok without it now?

**Comment by theuni** [comment]
Thanks for addressing my questions. utACK.

@laanwj I'd like to have a chance to look at the qt/gitian changes as well. Mind giving me until the end of this week to get to them? Ideally we'd get all of those in at once so everything remains in sync.

File: .travis.yml [hunk 11]
```diff
@@ -19,58 +19,63 @@ env:
...
-    - if [ -n "$DPKG_ADD_ARCH" ]; then sudo dpkg --add-architecture "$DPKG_ADD_ARCH" ; fi
-    - if [ -n "$PACKAGES" ]; then travis_retry sudo apt-get update; fi
-    - if [ -n "$PACKAGES" ]; then travis_retry sudo apt-get install --no-install-recommends --no-upgrade -qq $PACKAGES; fi
+    - env | grep -E '^(CCACHE_|WINEDEBUG|DISPLAY|BOOST_TEST_RANDOM)' | tee /tmp/env
+    - if [[ $HOST = *-mingw32 ]]; then DOCKER_ADMIN="--cap-add SYS_ADMIN"; fi
+    - DOCKER_ID=$(docker run $DOCKER_ADMIN -idt --mount type=bind,src=$TRAVIS_BUILD_DIR,dst=$TRAVIS_BUILD_DIR --mount type=bind,src=$CCACHE_DIR,dst=$CCACHE_DIR -w $TRAVIS_BUILD_DIR --env-file /tmp/env ubuntu:18.04)
+    - DOCKER_EXEC="docker exec $DOCKER_ID"
+    - if [ -n "$DPKG_ADD_ARCH" ]; then $DOCKER_EXEC dpkg --add-architecture "$DPKG_ADD_ARCH" ; fi
+    - travis_retry $DOCKER_EXEC apt update
+    - travis_retry $DOCKER_EXEC apt install --no-install-recommends --no-upgrade -qq -y $PACKAGES $DOCKER_PACKAGES
```
Comment (MarcoFalke) [inline .travis.yml]: I incorrectly assumed that apt is an alias for apt-get and asked you to change it. This was wrong, so feel free to change it back to apt-get. Sorry.

Comment (ken2812221) [reply .travis.yml]: Never mind

**Comment by ken2812221** [comment]
Update: also expose CONFIG_SHELL into docker

**Comment by jamesob** [comment]
re-utACK https://github.com/bitcoin/bitcoin/pull/13215/commits/67cfb4537ae7ca360f1b0cfb7e625aff9b59d8d7

File: .travis.yml [hunk 12]
```diff
@@ -19,58 +19,62 @@ env:
...
+    - if [ "$NEED_XVFB" = 1 ]; then $DOCKER_EXEC /sbin/start-stop-daemon --start --pidfile /tmp/custom_xvfb_99.pid --make-pidfile --background --exec /usr/bin/Xvfb -- :99 -ac; fi
 script:
     - export TRAVIS_COMMIT_LOG=`git log --format=fuller -1`
-    - if [ -n "$USE_SHELL" ]; then export CONFIG_SHELL="$USE_SHELL"; fi
     - OUTDIR=$BASE_OUTDIR/$TRAVIS_PULL_REQUEST/$TRAVIS_JOB_NUMBER-$HOST
     - BITCOIN_CONFIG_ALL="--disable-dependency-tracking --prefix=$TRAVIS_BUILD_DIR/depends/$HOST --bindir=$OUTDIR/bin --libdir=$OUTDIR/lib"
-    - if [ -z "$NO_DEPENDS" ]; then ccache --max-size=$CCACHE_SIZE; fi
-    - test -n "$USE_SHELL" && eval '"$USE_SHELL" -c "./autogen.sh"' || ./autogen.sh
+    - if [ -z "$NO_DEPENDS" ]; then $DOCKER_EXEC ccache --max-size=$CCACHE_SIZE; fi
+    - test -n "$USE_SHELL" && $DOCKER_EXEC "$USE_SHELL" -c "./autogen.sh" || $DOCKER_EXEC ./autogen.sh
```
Comment (MarcoFalke) [inline .travis.yml]: nit: Could reuse `CONFIG_SHELL` as below and skip exporting `USE_SHELL` or is there a reason to have two separate env vars with the same value?

**Comment by MarcoFalke** [comment]
re-ACK 67cfb4537a

File: .travis.yml (hunk 12 above)
Comment (ken2812221) [reply .travis.yml]: Done

**Review by jnewbery — COMMENTED** [review]
Concept ACK.

Build 4 here: https://travis-ci.org/bitcoin/bitcoin/jobs/380317076 took 36 minutes, which seems to be much longer than normal. I don't know whether that's caused by the changes in this PR or whether it's just natural variance.

**Comment by ken2812221** [comment]
@jnewbery Because I just modify the enviroment variable on that one and travis read caches based on enviroment variable. So it needs to recompile everything.

File: .travis.yml (hunk 10 above)
Comment (sipa) [reply .travis.yml]: I think we may not need this `travis_wait` anymore, due to several improvements to the unit tests' runtimes.

Comment (MarcoFalke) [reply .travis.yml]: @sipa is correct. See https://github.com/bitcoin/bitcoin/pull/12970#issuecomment-380952998 for background. The current run time (on the windows jobs) in this pull:

`i686-w64-mingw32`: 245.895077045 s
`x86_64-w64-mingw32`: 95.965838938 s

File: .travis.yml [hunk 13]
```diff
@@ -19,58 +19,62 @@ env:
...
-    - if [ -z "$NO_DEPENDS" ]; then ccache --max-size=$CCACHE_SIZE; fi
-    - test -n "$USE_SHELL" && eval '"$USE_SHELL" -c "./autogen.sh"' || ./autogen.sh
+    - if [ -z "$NO_DEPENDS" ]; then $DOCKER_EXEC ccache --max-size=$CCACHE_SIZE; fi
+    - test -n "$CONFIG_SHELL" && $DOCKER_EXEC "$CONFIG_SHELL" -c "./autogen.sh" || $DOCKER_EXEC ./autogen.sh
     - mkdir build && cd build
-    - ../configure --cache-file=config.cache $BITCOIN_CONFIG_ALL $BITCOIN_CONFIG || ( cat config.log && false)
-    - make distdir VERSION=$HOST
+    - $DOCKER_EXEC bash -c "cd build && ../configure --cache-file=config.cache $BITCOIN_CONFIG_ALL $BITCOIN_CONFIG" || ( cat config.log && false)
+    - $DOCKER_EXEC bash -c "cd build && make distdir VERSION=$HOST"
     - cd bitcoin-$HOST
```
Comment (jnewbery) [inline .travis.yml]: Can we remove this now?

Comment (MarcoFalke) [reply .travis.yml]: I think we mirror the pwd for the travis vm and the docker container. Otherwise a `cat config.log` on the travis vm wouldn't work, I presume.

Comment (MarcoFalke) [reply .travis.yml]: Oh yeah, it seems this one can be removed. Though, no harm in leaving it in, I guess.

Comment (ken2812221) [reply .travis.yml]: No, `cat config.log` below would not work then.

Comment (jnewbery) [reply .travis.yml]: ah, makes sense. Thanks

**Comment by ken2812221** [comment]
How about add `DOCKER_EXEC () { docker exec $DOCKER_ID bash -c "cd $PWD && $*"; }` as a bash funtion, so we don't need to `cd` all the time. See https://github.com/ken2812221/bitcoin/blob/483a525e8f09814366edcbf88803584a74ca1d2d/.travis.yml

**Comment by MarcoFalke** [comment]
re-utACK a182c9d

File: .travis.yml [hunk 14]
```diff
@@ -19,58 +19,62 @@ env:
...
+    - if [ "$NEED_XVFB" = 1 ]; then DOCKER_EXEC /sbin/start-stop-daemon --start --pidfile /tmp/custom_xvfb_99.pid --make-pidfile --background --exec /usr/bin/Xvfb -- :99 -ac; fi
 script:
     - export TRAVIS_COMMIT_LOG=`git log --format=fuller -1`
-    - if [ -n "$USE_SHELL" ]; then export CONFIG_SHELL="$USE_SHELL"; fi
     - OUTDIR=$BASE_OUTDIR/$TRAVIS_PULL_REQUEST/$TRAVIS_JOB_NUMBER-$HOST
     - BITCOIN_CONFIG_ALL="--disable-dependency-tracking --prefix=$TRAVIS_BUILD_DIR/depends/$HOST --bindir=$OUTDIR/bin --libdir=$OUTDIR/lib"
-    - if [ -z "$NO_DEPENDS" ]; then ccache --max-size=$CCACHE_SIZE; fi
-    - test -n "$USE_SHELL" && eval '"$USE_SHELL" -c "./autogen.sh"' || ./autogen.sh
+    - if [ -z "$NO_DEPENDS" ]; then DOCKER_EXEC ccache --max-size=$CCACHE_SIZE; fi
+    - test -n "$CONFIG_SHELL" && DOCKER_EXEC "CONFIG_SHELL" -c "./autogen.sh" || DOCKER_EXEC ./autogen.sh
```
Comment (MarcoFalke) [inline .travis.yml]: bash: CONFIG_SHELL: command not found

Comment (ken2812221) [reply .travis.yml]: Fixed.

**Comment by MarcoFalke** [comment]
re-utACK 59e9688eda4c4b01ee1713625632cd766c1a7ca9

**Review by jamesob — APPROVED** [review]
re-utACK https://github.com/bitcoin/bitcoin/pull/13215/commits/59e9688eda4c4b01ee1713625632cd766c1a7ca9

**Review by jamesob — APPROVED** [review]
re-utACK https://github.com/bitcoin/bitcoin/pull/13215/commits/59e9688eda4c4b01ee1713625632cd766c1a7ca9
